aeron-tools/src/main/java/uk/co/real_logic/aeron/tools/SubscriberTool.java

```diff
@@ -0,0 +1,719 @@
+        {
+            totalMessages += subscribers[i].verifiableMessagesReceived();
+        }
+        return totalMessages;
+    }
+
+    /**
+     * A snapshot (non-atomic) total of the number of bytes
+     * received across all receiving threads.
+     * @return current total number of bytes received
+     */
+    @Override
+    public long bytes()
+    {
+        long totalBytes = 0;
+        for (int i = 0; i < subscribers.length; i++)
+        {
+            totalBytes += subscribers[i].bytesReceived();
+        }
+        return totalBytes;
+    }
+
+    /**
+     * A snapshot (non-atomic) total of the number of non-verifiable messages
+     * received across all receiving threads.
+     * @return current total number of non-verifiable messages received
+     */
+    @Override
+    public long nonVerifiableMessages()
+    {
+        long totalMessages = 0;
+        for (int i = 0; i < subscribers.length; i++)
+        {
+            totalMessages += subscribers[i].nonVerifiableMessagesReceived();
+        }
+        return totalMessages;
+    }
+
+    /**
+     * Optionally sets the random seed used for the TLRandom class, and reports the seed used.
+     * @return the seed to use
+     */
+    @Override
+    public long setSeed(long seed)
+    {
+        if (options.getRandomSeed() != 0)
+        {
+            seed = options.getRandomSeed();
+        }
+        LOG.info("{}", String.format("Thread %s using random seed %d.", Thread.currentThread().getName(), seed));
+        return seed;
+    }
+
+    class SubscriberThread implements Runnable, InactiveConnectionHandler, NewConnectionHandler, RateController.Callback
+    {
+        final int threadId;
+        private long nonVerifiableMessagesReceived;
+        private long verifiableMessagesReceived;
+        private long bytesReceived;
+        private byte[] bytesToWrite = new byte[1];
+        private final Aeron.Context ctx;
+        private final Aeron aeron;
+        private final UnsafeBuffer controlBuffer = new UnsafeBuffer(new byte[4 + 4 + CHANNEL_NAME_MAX_LEN + 4 + 4]);
+        private final RateController rateController;
+
+        /* All subscriptions we're interested in. */
+        final Subscription subscriptions[];
+        /* Just the subscriptions we have reason to believe might have data available - these we actually poll on. */
+        final Subscription activeSubscriptions[];
+        int activeSubscriptionsLength = 0;
+
+        private final Publication controlPublication;
+        private final Subscription controlSubscription;
+        /* Doesn't use TLRandom, since this really does need to be random and shouldn't
+         * be affected by manually setting the seed. */
+        private final int controlSessionId = ThreadLocalRandom.current().nextInt();
+
+        /* channel -> stream ID -> session ID */
+        private final HashMap<String, Int2ObjectHashMap<Int2ObjectHashMap<MessageStream>>> messageStreams =
```

Comment:
I also noticed subscriber tool has some “incorrect” whitespace formatting. Want me to clean that up?

> On May 4, 2015, at 2:51 PM, Martin Thompson <notifications@github.com> wrote:
> 
> In aeron-tools/src/main/java/uk/co/real_logic/aeron/tools/SubscriberTool.java <https://github.com/real-logic/Aeron/pull/108#discussion_r29616645>:
> 
> > +        private final RateController rateController;
> > +
> > +        /* All subscriptions we're interested in. */
> > +        final Subscription subscriptions[];
> > +        /* Just the subscriptions we have reason to believe might have data available - these we actually poll on. */
> > +        final Subscription activeSubscriptions[];
> > +        int activeSubscriptionsLength = 0;
> > +
> > +        private final Publication controlPublication;
> > +        private final Subscription controlSubscription;
> > +        /* Doesn't use TLRandom, since this really does need to be random and shouldn't
> > +         * be affected by manually setting the seed. */
> > +        private final int controlSessionId = ThreadLocalRandom.current().nextInt();
> > +
> > +        /* channel -> stream ID -> session ID */
> > +        private final HashMap<String, Int2ObjectHashMap<Int2ObjectHashMap<MessageStream>>> messageStreams =
> I'd ask a different question. Why do you need comments on normal fields? It is always a smell to me when things need comments. Over time the code changes and the comment become misleading and even plain wrong.
> 
> If the code is so unclear then how will a comment help? If the person cannot write clear enough code that it requires comments then what makes you think the comments are going to be better than then code?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/real-logic/Aeron/pull/108/files#r29616645>.
>